src/internal/pachsql/url.go

```diff
@@ -0,0 +1,104 @@
+package pachsql
+
+import (
+	"net"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/go-sql-driver/mysql"
+	"github.com/jmoiron/sqlx"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+)
+
+// URL contains the information needed to connect to a SQL database, except for the password.
+type URL struct {
+	Protocol string
+	User     string
+	Host     string
+	Port     uint16
+	Database string
+	Params   map[string]string
+}
+
+// ParseURL attempts to parse x into a URL
+func ParseURL(x string) (*URL, error) {
+	u, err := url.Parse(x)
+	if err != nil {
+		return nil, err
+	}
+	port, err := strconv.Atoi(u.Port())
+	if err != nil {
+		return nil, err
+	}
+	params := make(map[string]string)
+	for k, v := range u.Query() {
+		if len(v) > 0 {
+			params[k] = v[len(v)-1]
```

Comment:
do we already know what param keys may have multiple values?

Reply:
I don't think any of them will have multiple values.  A URL is allowed have multiple values per parameter though.